File: DataStructure_TA_PondS/Data Structure Algorithm/Source Code/Chapter 4/C/DoubleLinkList.cpp
#include <stdio.h>
#include <DoubleLink.h>
#define true 1
#define false 0
typedef int boolean;
struct DoubleLink *newNode;
struct DoubleLink *head;
struct DoubleLink *last;
struct DoubleLink *curr;
struct DoubleLink *prev1;
struct DoubleLink *prev2;

void add(int newItem) {
  if (head == NULL) {
    newNode = insertNode(newItem, NULL, NULL);
    last = newNode;
  } else {
    newNode = insertNode(newItem, head, NULL);
    head->precede = newNode;
    head = newNode;
  }
  head = newNode;
}

boolean searchItem(int item) {
  curr = head;
  prev1 = NULL;
  boolean status = false;
  while (curr != NULL) {
    if (curr->item == item) {
      prev2 = curr->next;
      status = true;
      break;
    } else {
      prev1 = curr;
      curr = curr->next;
    }
 }
 if (status) {
   return true;
 } else {
   return false;
 }
}

void deleteNode(int item) {
  if (searchItem(item)) {
    if (prev1 == NULL) {
      head = curr->next;
      if (prev2 != NULL) {
        prev2->precede = NULL;
      }
    } else if (prev2 == NULL) {
      prev1->next = NULL;
      curr->precede = NULL;
      last = prev1;
    } else {
      prev1->next = curr->next;
      prev2->precede = curr->precede;
    }
  } else {
    printf("Not found data");
  }
}

void insert(int iteminsert, int newItem) {
  newNode = insertNode(newItem, NULL, NULL);
  if (searchItem(iteminsert)) {
    prev2 = curr;
    if (prev1 == NULL) {
      newNode->next = curr;
      curr->precede = newNode;
      head = newNode;
    } else {
      newNode->next = prev1->next;
      prev1->next = newNode;
      newNode->precede = prev2->precede;
      prev2->precede = newNode;
    }
  } else {
     if (head == NULL) {
       newNode = insertNode(newItem,NULL,NULL);
      last = newNode;
    } else {
      newNode = insertNode(newItem,NULL,last);
      last->next = newNode;
      last = newNode;
    }
  }
}

void showdata() {
  curr = head;
  while (curr != NULL) {
    printf("%d ",curr->item);
    curr = curr->next;
  }
  if (head != NULL) {
    printf(" backward : ");
    curr = last;
    while (curr != NULL) {
      printf("%d ",curr->item);
      curr = curr->precede;
    }
    printf("\n");
  }
}

void main() {
  add(9);
  printf("add 9  : ");
  showdata();
  add(10);
  printf("add 10 : ");
  showdata();
  add(11);
  printf("add 11 : ");
  showdata();
  add(12);
  printf("add 12 : ");
  showdata();
  deleteNode(10);
  printf("delete  10 : ");
  showdata();
  deleteNode(12);
  printf("delete  12 : ");
  showdata();
  insert(11, 15);
  printf("insert 15 before 11 : ");
  showdata();
  insert(11, 17);
  printf("insert 17 before 11 : ");
  showdata();
  insert(18, 19);
  printf("insert 19 at last Link List : ");
  showdata();
  scanf("%d");
}

File: DataStructure_TA_PondS/Data Structure Algorithm/Source Code/Chapter 4/Java/Chapter4/src/chapter4/DoubleLinkList.java
package chapter4;

public class DoubleLinkList {
  DoubleLink newNode;
  DoubleLink head;
  DoubleLink last;
  DoubleLink curr;
  DoubleLink prev1;
  DoubleLink prev2;
  //เพิ่มข้อมูลในลิงค์ลิสต์แบบสองทิศทาง
  public void add(Object newItem) {
    if (head == null) {
      newNode = new DoubleLink(newItem);
      last = newNode;
    } else {
      newNode = new DoubleLink(newItem, head, null);
      head.setPrecede(newNode);             
      head = newNode;
    }
    head = newNode;
  }

  public boolean searchItem(Object item) {
    curr = head;
    prev1 = null;
    boolean status = false;
    while (curr != null) {
      if (curr.getItem() == item) {
        prev2 = curr.getNext();
        status = true;
        break;
      } else {
        prev1 = curr;
        curr = curr.getNext();
      }
    }

    if (status) {
      return true;
    } else {
      return false;
    }
  }

  public void deleteNode(Object item) {
    if (searchItem(item)) {
      if (prev1 == null) {
        head = curr.getNext();
        if (prev2 != null) {
          prev2.setPrecede(null);
        }
      } else if (prev2 == null) {
        prev1.setNext(null);
        curr.setPrecede(null);
        last = prev1;
      } else {
        prev1.setNext(curr.getNext());
        prev2.setPrecede(curr.getPrecede());
      }
    } else {
      System.out.println("Not found data");
    }
  }

  public void insert(Object iteminsert, Object newItem) {
    newNode = new DoubleLink(newItem);
    if (searchItem(iteminsert)) {
      prev2 = curr;
      if (prev1 == null) {
        newNode.setNext(curr);
        curr.setPrecede(newNode);
        head = newNode;
      } else {
        newNode.setNext(prev1.getNext());
        prev1.setNext(newNode);
        newNode.setPrecede(prev2.getPrecede());
        prev2.setPrecede(newNode);
      }
    } else {
      if (head == null) {
        newNode = new DoubleLink(newItem);
        last = newNode;
      } else {
        newNode = new DoubleLink( newItem,  null, last);
        last.setNext(newNode);
        last = newNode;
      }
    }
  }

  public void showdata() {
    curr = head;
    while (curr != null) {
      System.out.print(curr.getItem()+" ");
      curr = curr.getNext();
    }
    if (head != null) {
      System.out.print(" backward : ");
      curr = last;
      while (curr != null) {
        System.out.print(curr.getItem()+" ");
        curr = curr.getPrecede();
      }
      System.out.println();
    }
  }

  public static void main(String[] args) {
    DoubleLinkList linklist = new DoubleLinkList();
    linklist.add(9);
    System.out.print("add 9  : ");
    linklist.showdata();
    linklist.add(10);
    System.out.print("add 10 : ");
    linklist.showdata();
    linklist.add(11);
    System.out.print("add 11 : ");
    linklist.showdata();
    linklist.add(12);
    System.out.print("add 12 : ");
    linklist.showdata();
    linklist.deleteNode(10);
    System.out.print("delete 10 : ");
    linklist.showdata();
    linklist.deleteNode(12);
    System.out.print("delete 12 : ");
    linklist.showdata();
    linklist.insert(11, 15);
    System.out.print("insert 15 before 11:");
    linklist.showdata();
    linklist.insert(11, 17);
    System.out.print("insert 17 before 11:");
    linklist.showdata();
    linklist.insert(18, 19);
    System.out.print("insert 19 at last : ");
    linklist.showdata();
  }    
}
